### ISSUE TITLE
Inconsistent Threshold of RigidBody2D Collision Responses

### ISSUE THREAD

**Issue by RationalFragile** [godotengine/godot#50595]
### Godot version

3.3.2.stable.official

### System information

Windows 10, GLES3

### Issue description

A RigidBody2D (Character mode) will get stuck between two StaticBody2D objects because the overlap distance is inconsistent depending on the speed at which the rigid body hit the static body.

Here is the setup (example project attached):
![RigidBody2D Stuck between Static bodies](https://user-images.githubusercontent.com/83667624/126081996-f70e1768-0d49-4de4-a136-d8849dda9e62.png)

Here is a diagram of the problematic corner:
![Collision Normals](https://user-images.githubusercontent.com/83667624/126082009-acc78be2-4763-4781-af1b-58bd4f31980a.png)

When the blue rigid body hits the green static body with any sufficient velocity (whether gravity is on or off), it will always settle at 0.3 overlap on the y axis. But if the blue hits at a low velocity, it settles at lower values (like 0.08).
The issue is that while the rigid body is not moving at all (and no gravity or downward forces exist) it will still stay at 0.3 overlap with the green, but cannot go deeper than 0.2 for example in the x axis to overlap the red.
This results in the case where the overlap rectangle between blue and red is taller than it is wide, which generates a horizontal contact normal (which makes sense) but means that the rigid body cannot move over the seam even though everything is snapped to pixels/grid.

In other words, the overlap displacement threshold (at which two overlapping bodies are considered just touching and not overlapping) should be the same *at rest* regardless of how fast the collision was.

Also, is there an easy way to modify this value of 0.3 max overlap? Tried with the custom solver bias of the CollisionShape2D but it had no effect.

Thank you a lot for reading my issue!

### Steps to reproduce

1. Open the attached project and start the only scene
2. Wait for the blue RigidBody2D to fall (using gravity, or disable gravity and press down arrow)
3. Wait for the blue to hit the green StaticBody2D, you should see that the y label offset shows 0.30000
4. Now press the right arrow (or left) and the blue will be stopped by the red static body and the x offset will be less than 0.3

### Minimal reproduction project

[Collision2DTest.zip](https://github.com/godotengine/godot/files/6836969/Collision2DTest.zip)

**Comment by RationalFragile** [comment]
And I forgot to mention that the continuous mode (discrete, ray, shape) does not affect the behavior I'm describing, neither does switching from Character mode to Rigid body mode.

**Comment by pouleyKetchoupp** [comment]
I haven't investigated your specific case yet, but could you try to set this parameter in the physics server in order to reduce the allowed penetration?

You can put this script on any node in the tree, just called once in _ready (it affects all rigid bodies).

```
var space = get_world_2d().space
Physics2DServer.space_set_param(space, Physics2DServer.SPACE_PARAM_BODY_MAX_ALLOWED_PENETRATION, 0.0)
```

In this example, it will be set to 0 but you can set any other value.

**Comment by RationalFragile** [comment]
Thank you for your quick reply.

After some testing, here is how I can summarize what I found:
- there is a critical velocity below which the seam will be able to stop the rigid body. The less the SPACE_PARAM_BODY_MAX_ALLOWED_PENETRATION (MAP) is, the less that critical velocity will be.
- If you set the MAP to zero, if there is gravity (or any downward force), then the critical velocity will still be higher than zero (albeit small, like 0.1 tiles per second).
- Otherwise if there are no downward forces and MAP is zero, then this critical velocity drops to almost zero (floating point errors).

I assume that the first and second cases are an issue because the seam shouldn't stop the rigid body even when it is pushed down (zero damping, zero friction). I'm not sure if this is an inherent limitation with the collision response algorithm or something can be done to account for corners/seams and slow moving objects.

But overall, setting the max overlap to zero does effectively help by making the seam only affect much slower moving objects. But, it creates another issue; very noticeable jitter at rest if you set a high gravity scale for example. Basically, every half a second, the y offset jumps to 0.1 or so, and goes back to zero. (And again, continuous collision modes have no effect on this jitter.)

**Comment by pouleyKetchoupp** [comment]
Thanks for the tests and detailed explanation!

After testing, I can confirm there's a bug here which is similar to #47148. There should be a better mechanism to prevent collision with edges between objects.

I'm closing this issue as duplicate of #47148, but this repro case will be useful to help solving the issue so I've added a note there about your findings.

**Comment by RationalFragile** [comment]
After spending the whole day reading the code, I figured something out, so trying to share what I found if it helps someone.

In the file `godot-3.3.2-stable\servers\physics_2d\collision_solver_2d_sat.cpp`

        (...)
        _FORCE_INLINE_ void generate_contacts() {

		// nothing to do, don't generate
		if (best_axis == Vector2(0.0, 0.0))
			return;

		// ---- START OF CHANGE ----
		real_t max_penetration = 0.3;
		if (best_depth <= max_penetration)
			return;
            // ---- END OF CHANGE ----

        (...)

This solves the issue (assuming you also set `Physics2DServer.SPACE_PARAM_BODY_MAX_ALLOWED_PENETRATION` to the same value). Sorry I couldn't find how to access the body/area from the shape to retrieve the value of the space using Physics2DServer space_get_param... Maybe someone who spent more time with the code could save me the hassle of figuring that out :D

Also, there is something I should correct about what I said earlier: the jittering issue seems to be completely unrelated, it happens even if the max_penetration is not zero. It too will happen at a certain point if the force pushing an object against another is sufficiently high. So basically, gravity scale set to 10 will jitter even with 0.3 max_penetration. And the jittering doesn't happen every frame, more like every half a second, so it must be some cached value (object set to never sleep); I played with the max_separation but it didn't help.

Anyway, thanks again!
